Add github_webhook module

stpierre opened the pull request:
##### SUMMARY

`github_webhook` lets you easily create, update and delete Github webhooks.

This is part of a two-part replacement for `github_hooks`, which lacks a number of important features and has a strange interface.

The other parts are:

* Add `github_webhook_facts` module: https://github.com/ansible/ansible/pull/35814
* Deprecate `github_hooks` module: #49296 

`github_hooks` lacks a number of important features: The ability to specify the events on which a webhook should fire; the ability to add a shared secret to a hook; the ability to create hooks that do not verify SSL against their target; and more. I considered merely updating that module to add the features needed, but the existing interface is pretty odd and would be difficult to update in a backwards-compatible way. For instance, it has a built-in feature for deleting all hooks whose last request returned 504 (`action: clean504`), but no way to update an existing hook, or delete a specific hook. It requires you to provide the API URL to a repository, rather than just the name of the repo (and optionally a Github API base), which requires a fair degree of familiarity with the Github API.

It has other issues that are easier to fix, but still weighed in on my decision to replace rather than upgrade: it rolls its own API client layer rather than using an existing library; it appears to support both password and API token auth, but only accepts a single `oauthkey` argument, which is a misnomer in both cases.

Rather than try to update it and fix all of the issues, I replaced it with two modules whose usage patterns are, I hope, much more intuitive.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

github_webhook

##### ANSIBLE VERSION
```
ansible 2.5.0 (github-webhook-module 4cdc321bfb) last updated 2018/02/06 16:13:32 (GMT -500)
  config file = None
  configured module search path = [u'/home/stpierre/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/stpierre/devel/ansible/lib/ansible
  executable location = /home/stpierre/devel/ansible/bin/ansible
  python version = 2.7.14 (default, Jan 17 2018, 14:28:32) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

djmattyg007 commented:
shipit

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+#!/usr/bin/python
+#
+# Copyright: (c) Ansible Project
```
```
# Copyright: (c) 2018, Ansible Project
```

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+DOCUMENTATION = '''
+---
+module: github_webhook
+short_description: Manage Github webhooks
```
s/Github/GitHub/

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+short_description: Manage Github webhooks
+version_added: "2.6"
+description:
+  - "Create and delete Github webhooks"
```
Can you please add more description here?

stpierre replied:
What more is needed? This seems sufficient to me.

stpierre replied:
What more is needed? This seems sufficient to me.

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+description:
+  - "Create and delete Github webhooks"
+requirements:
+  - "PyGithub >= 1.3.5"
```
Can you please use [`github3.py`](https://github.com/sigmavirus24/github3.py) instead of `PyGithub` as other github_* modules uses that library?

stpierre replied:
Not really, no. I tested it a bit, and `github3.py` seems to have pretty dismal support for GitHub Enterprise, which is a critical feature for me.

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+requirements:
+  - "PyGithub >= 1.3.5"
+options:
+  repository:
```
add aliases like 

```
aliases: ['repo']
```

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+    description:
+      - The shared secret between Github and the payload URL.
+    required: false
+  insecure_ssl:
```
Can you rename this to `validate_certs` to match Ansible parameter naming convention ?

stpierre replied:
I'd really rather not, for two reasons:

1. That introduces an asymmetricity between the inputs and outputs of the webhook modules. See the previous discussion on this PR for details.

2. This doesn't do the same thing as other `validate_certs` options. I've clarified the documentation.

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+    required: true
+  url:
+    description:
+      - URL to which payloads will be delivered
```
Add description that this different that `github_url` parameter ?

stpierre replied:
I feel like this is pretty obvious. I'm not sure how to clarify it beyond the descriptions that are already in place.

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+        module.params["user"],
+        module.params.get("password") or module.params.get("token"),
+        base_url=module.params["github_url"])
+
```
Add check here to see if user provided password or token actually works or not ?

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+                break
+    except github.GithubException as err:
+        module.fail_json(
+            msg="Unable to get hooks from repository %s: %s" % to_native(err),
```
Too few format values 

```
Unable to get hooks from repository : %s" % to_native(err)
```

Akasurde commented on lib/ansible/modules/source_control/github_webhook.py:
```diff
@@ -0,0 +1,240 @@
+        changed, data = create_hook(repo, module)
+    elif hook is not None and module.params["state"] == "absent":
+        hook.delete()
+        changed = True
```
This will always return changed even if there is failure in deleting hook.

stpierre committed:
Add github_webhook module

github_webhook lets you easily create, update and delete Github
webhooks.

This is part of a two-part replacement for github_hooks, which lacks a
number of important features and has a strange interface.

stpierre commented:
@Akasurde I believe your objections have mostly been addressed.

Akasurde commented:
Need another pair of eyes to approve this.

acozine commented:
@stpierre thanks for the contributions. Are you willing to maintain this module and the `github_webhook_facts` module? If so, please add yourself to the maintenance list (above [this line](https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml#L315)).

gundalow committed:
use `-name:` in EXAMPLES

gundalow commented:
I made a change by specifying `-name: ` for the examples https://github.com/ansible/ansible/pull/35813/commits/d667575c807e851048d973ff1c7dbfc86ee359e7

gundalow committed:
typo
